### ISSUE TITLE
注意力实现中掩码矩阵是不是没考虑past_kv的长度

### ISSUE THREAD

**Issue by CZH-THU** [jingyaogong/minimind#631]
model_minimind.py 198-121行：
```
        else:
            scores = (xq @ xk.transpose(-2, -1)) / math.sqrt(self.head_dim)
            scores = scores + torch.triu(
                torch.full((seq_len, seq_len), float("-inf"), device=scores.device),
                diagonal=1
            ).unsqueeze(0).unsqueeze(0)  # scores+mask

            if attention_mask is not None:
                extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
                extended_attention_mask = (1.0 - extended_attention_mask) * -1e9
                scores = scores + extended_attention_mask

            scores = F.softmax(scores.float(), dim=-1).type_as(xq)
            scores = self.attn_dropout(scores)
            output = scores @ xv
```
这部分的注意力实现 掩码矩阵是不是没考虑past_kv的长度，仅考虑当前seq_len？这样做没问题么

**Comment by jingyaogong** [comment]
kv_cache 只在自回归推理使用，此时 seq_len=1，causal mask triu((1,1), diagonal=1) 结果是空矩阵（不遮挡任何位置），这就是预期行为，新 token attend 到所有历史 token。

---

训练不用 kv_cache，seq_len 等于完整序列长度，mask 是标准 (seq_len, seq_len) 下三角矩阵。

**Comment by CZH-THU** [comment]
> kv_cache 只在自回归推理使用，此时 seq_len=1，causal mask triu((1,1), diagonal=1) 结果是空矩阵（不遮挡任何位置），这就是预期行为，新 token attend 到所有历史 token。
> 
> 训练不用 kv_cache，seq_len 等于完整序列长度，mask 是标准 (seq_len, seq_len) 下三角矩阵。

在与模型的多轮会话时，用户的输入token 长度超过1，如果我想实现提示词缓存的功能即复用之前会话历史的kvcache，代码的实现是不是就cover不住了

**Comment by jingyaogong** [comment]
”代码的实现是不是就cover不住了“ 

复用前缀kv cache那这边把后续的seq_len加上mask就行了，实现比较容易，拉最新代码：https://github.com/jingyaogong/minimind/commit/1279a6168171c859d70d46727975eb541de4fa55

MiniMindForCausalLM最后增加：

https://github.com/jingyaogong/minimind/issues/631#issuecomment-3744288529 已编辑

```python
    @torch.inference_mode() 
    def generate(self, inputs, attention_mask=None, max_new_tokens=8192, temperature=0.85, top_p=0.85, top_k=50, eos_token_id=None, streamer=None, use_cache=True, num_return_sequences=1, do_sample=True, repetition_penalty=1.0, **kwargs):
        input_ids = kwargs.get("input_ids", inputs).repeat(num_return_sequences, 1)
        attention_mask = attention_mask.repeat(num_return_sequences, 1) if attention_mask is not None else None
        past_key_values = kwargs.get("past_key_values", None)
        if streamer: streamer.put(input_ids.cpu())
        for _ in range(max_new_tokens):
            past_len = past_key_values[0][0].shape[1] if past_key_values else 0
            outputs = self.forward(input_ids[:, past_len:], attention_mask, past_key_values, use_cache=use_cache)
            attention_mask = torch.cat([attention_mask, attention_mask.new_ones(attention_mask.shape[0], 1)], -1) if attention_mask is not None else None
            logits = outputs.logits[:, -1, :] / temperature
            if repetition_penalty != 1.0:
                for i in range(input_ids.shape[0]): logits[i, torch.unique(input_ids[i])] /= repetition_penalty
            if top_k > 0: 
                logits[logits < torch.topk(logits, top_k)[0][..., -1, None]] = -float('inf')
            if top_p < 1.0:
                sorted_logits, sorted_indices = torch.sort(logits, descending=True)
                mask = torch.cumsum(torch.softmax(sorted_logits, dim=-1), dim=-1) > top_p
                mask[..., 1:], mask[..., 0] = mask[..., :-1].clone(), 0
                logits[mask.scatter(1, sorted_indices, mask)] = -float('inf')
            next_token = torch.multinomial(torch.softmax(logits, dim=-1), num_samples=1) if do_sample else torch.argmax(logits, dim=-1, keepdim=True)
            input_ids = torch.cat([input_ids, next_token], dim=-1)
            past_key_values = outputs.past_key_values if use_cache else None
            if streamer: streamer.put(next_token.cpu())
            if eos_token_id and (next_token == eos_token_id).any(): break
        if streamer: streamer.end()
        if kwargs.get("return_kv"): return {'generated_ids': input_ids, 'past_kv': past_key_values}
        return input_ids
```

eval_llm.py
https://github.com/jingyaogong/minimind/issues/631#issuecomment-3744288529 已编辑
```python
def main():
    parser = argparse.ArgumentParser(description="MiniMind模型推理与对话")
    parser.add_argument('--load_from', default='model', type=str, help="模型加载路径（model=原生torch权重，其他路径=transformers格式）")
    parser.add_argument('--save_dir', default='out', type=str, help="模型权重目录")
    parser.add_argument('--weight', default='full_sft', type=str, help="权重名称前缀（pretrain, full_sft, rlhf, reason, ppo_actor, grpo, spo）")
    parser.add_argument('--lora_weight', default='None', type=str, help="LoRA权重名称（None表示不使用，可选：lora_identity, lora_medical）")
    parser.add_argument('--hidden_size', default=768, type=int, help="隐藏层维度（512=Small-26M, 640=MoE-145M, 768=Base-104M）")
    parser.add_argument('--num_hidden_layers', default=8, type=int, help="隐藏层数量（Small/MoE=8, Base=16）")
    parser.add_argument('--use_moe', default=0, type=int, choices=[0, 1], help="是否使用MoE架构（0=否，1=是）")
    parser.add_argument('--inference_rope_scaling', default=False, action='store_true', help="启用RoPE位置编码外推（4倍，仅解决位置编码问题）")
    parser.add_argument('--max_new_tokens', default=8192, type=int, help="最大生成长度（注意：并非模型实际长文本能力）")
    parser.add_argument('--temperature', default=0.85, type=float, help="生成温度，控制随机性（0-1，越大越随机）")
    parser.add_argument('--top_p', default=0.85, type=float, help="nucleus采样阈值（0-1）")
    parser.add_argument('--historys', default=20, type=int, help="携带历史对话轮数（需为偶数，0表示不携带历史）")
    parser.add_argument('--show_speed', default=1, type=int, help="显示decode速度（tokens/s）")
    parser.add_argument('--device', default='cuda' if torch.cuda.is_available() else 'cpu', type=str, help="运行设备")
    args = parser.parse_args()
    
    prompts = [
        '你有什么特长？',
        '为什么天空是蓝色的',
        '请用Python写一个计算斐波那契数列的函数',
        '解释一下"光合作用"的基本过程',
        '如果明天下雨，我应该如何出门',
        '比较一下猫和狗作为宠物的优缺点',
        '解释什么是机器学习',
        '推荐一些中国的美食'
    ]
    
    conversation, past_kv = [], None
    model, tokenizer = init_model(args)
    input_mode = int(input('[0] 自动测试\n[1] 手动输入\n'))
    streamer = TextStreamer(tokenizer, skip_prompt=True, skip_special_tokens=True)
    
    prompt_iter = prompts if input_mode == 0 else iter(lambda: input('💬: '), '')
    for idx, prompt in enumerate(prompt_iter):
        setup_seed(2026) # or setup_seed(random.randint(0, 2048))
        if input_mode == 0: print(f'💬: {prompt}')
        if not args.historys:
            conversation, past_kv = [], None
        elif args.historys < 9999 and len(conversation) >= args.historys * 2:
            conversation, past_kv = conversation[-(args.historys - 1) * 2:], None
        conversation.append({"role": "user", "content": prompt})

        templates = {"conversation": conversation, "tokenize": False, "add_generation_prompt": True}
        if args.weight == 'reason': templates["enable_thinking"] = True # 仅Reason模型使用
        inputs = tokenizer.apply_chat_template(**templates) if args.weight != 'pretrain' else (tokenizer.bos_token + prompt)
        # 无padding
        # inputs = tokenizer(inputs, return_tensors="pt", truncation=True).to(args.device)
        # 左padding
        inputs = tokenizer(inputs, return_tensors="pt", truncation=True, max_length=200+int(idx*500), padding_side='left', padding="max_length").to(args.device)

        print('🤖: ', end='')
        st = time.time()
        outputs = model.generate(
            inputs=inputs["input_ids"], attention_mask=inputs["attention_mask"],
            max_new_tokens=args.max_new_tokens, do_sample=True, streamer=streamer,
            pad_token_id=tokenizer.pad_token_id, eos_token_id=tokenizer.eos_token_id,
            top_p=args.top_p, temperature=args.temperature, repetition_penalty=1.0,
            past_key_values=past_kv, return_kv=True
        )
        generated_ids, past_kv = outputs['generated_ids'], outputs['past_kv']
        response = tokenizer.decode(generated_ids[0][len(inputs["input_ids"][0]):], skip_special_tokens=True)
        conversation.append({"role": "assistant", "content": response})
        gen_tokens = len(generated_ids[0]) - len(inputs["input_ids"][0])
        print(f'\n[Speed]: {gen_tokens / (time.time() - st):.2f} tokens/s\n\n') if args.show_speed else print('\n\n')
```

即可实现多轮对话复用前轮prefill前缀kv cache

**Comment by CZH-THU** [comment]
@jingyaogong 感谢回复，看了下代码，还有两个疑问：
1、generate 函数中 是不是要加个类似
```
                attention_mask = torch.cat(
                    [attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1
                )
```
这样的逻辑更新一下attention mask
2、     
 ```
      if past_len:
                outputs = self.forward(input_ids[:, past_len:], None, past_key_values, use_cache=use_cache)
```
这块为啥不传attention_mask 呢？ past kv 里面有可能有padding token，个人认为新token计算时需要遮蔽掉这部分注意力

**Comment by jingyaogong** [comment]
> [@jingyaogong](https://github.com/jingyaogong) 感谢回复，看了下代码，还有两个疑问： 1、generate 函数中 是不是要加个类似
> 
> ```
>                 attention_mask = torch.cat(
>                     [attention_mask, attention_mask.new_ones((attention_mask.shape[0], 1))], dim=-1
>                 )
> ```
> 
> 这样的逻辑更新一下attention mask 2、
> 
> ```
>      if past_len:
>                outputs = self.forward(input_ids[:, past_len:], None, past_key_values, use_cache=use_cache)
> ```
> 
> 这块为啥不传attention_mask 呢？ past kv 里面有可能有padding token，个人认为新token计算时需要遮蔽掉这部分注意力

这两点都是正确的

主要取决于要不要做left pad，现有的eval_llm配套是不需要attention_mask的，那就可以留空

完全版本上面comment已经更新： https://github.com/jingyaogong/minimind/issues/631#issuecomment-3743287130
